static/static_file_modifier.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2015 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package static provides a modifier that allow Martian to reurn static files
+// local to Martian.
+package static
+
+import (
+	"encoding/json"
+	"mime"
+	"net/http"
+	"os"
+	"path"
+	"path/filepath"
+
+	"github.com/google/martian"
+	"github.com/google/martian/parse"
+)
+
+type staticModifier struct {
+	rootPath string
+}
+
+type staticJSON struct {
+	RootPath string               `json:"rootPath"`
+	Scope    []parse.ModifierType `json:"scope"`
+}
+
+func init() {
+	parse.Register("static.Modifier", modifierFromJSON)
+}
+
+// NewModifier constructs a staticModifier that takes a path from which to
+// serve files from.
+func NewModifier(rootPath string) martian.RequestResponseModifier {
+	return &staticModifier{
+		rootPath: path.Clean(rootPath),
+	}
+}
+
+// ModifyRequest marks the context to skip the roundtrip.
+func (s *staticModifier) ModifyRequest(req *http.Request) error {
+	ctx := martian.NewContext(req)
+	ctx.SkipRoundTrip()
+	return nil
+}
+
+// ModifyResponse reads a the file rooted at rootPath joined with the request URL
+// path.  In the case that the file cannot be found, the response will be a 404.
+func (s *staticModifier) ModifyResponse(res *http.Response) error {
+	p := filepath.Join(s.rootPath, filepath.Clean(res.Request.URL.Path))
+	f, err := os.Open(p)
+	switch {
+	case os.IsNotExist(err):
+		res.StatusCode = 404
+		return err
+	case os.IsPermission(err):
```

Comment:
Does this fail in the case of permissions of the caller being insufficient or the martian proxy itself?  If it's the proxy itself, 401 seems slightly misleading to me since it would indicate that the request itself needs different authorization, i.e. that it could possibly be fixed by changing the request--but that's the not the case.  Maybe 500 would be more apt?